File: tests/TestAgentActions.cpp
#include <QTest>
#include <QSignalSpy>
#include <QVariantMap>
#include "SignalBus.h"

class TestAgentActions : public QObject {
    Q_OBJECT

private slots:
    void testAuthActions() {
        SignalBus bus;
        QSignalSpy spy(&bus, &SignalBus::agentUiActionRequested);

        QVariantMap params;
        params["username"] = "Admin";
        params["password"] = "1";
        emit bus.agentUiActionRequested("admin.login", params);

        QCOMPARE(spy.count(), 1);
        QList<QVariant> args = spy.takeFirst();
        QCOMPARE(args.at(0).toString(), QString("admin.login"));
        QCOMPARE(args.at(1).toMap()["username"].toString(), QString("Admin"));
        QCOMPARE(args.at(1).toMap()["password"].toString(), QString("1"));

        emit bus.agentUiActionRequested("admin.logout", QVariantMap());
        QCOMPARE(spy.count(), 1);
        QCOMPARE(spy.takeFirst().at(0).toString(), QString("admin.logout"));
    }

    void testAssistantActions() {
        SignalBus bus;
        QSignalSpy spy(&bus, &SignalBus::agentUiActionRequested);
        
        QStringList actions = {
            "assistant.open", "assistant.close",
            "assistant.reload_model", "assistant.reload_rag",
            "assistant.reload_agent", "assistant.reload_server"
        };
        
        for (const QString& action : actions) {
            emit bus.agentUiActionRequested(action, QVariantMap());
            QCOMPARE(spy.count(), 1);
            QCOMPARE(spy.takeFirst().at(0).toString(), action);
        }
    }

    void testMailActions() {
        SignalBus bus;
        QSignalSpy spy(&bus, &SignalBus::agentUiActionRequested);
        
        QStringList actions = { "mail.open", "mail.close", "mail.settings" };
        for (const QString& action : actions) {
            emit bus.agentUiActionRequested(action, QVariantMap());
            QCOMPARE(spy.count(), 1);
            QCOMPARE(spy.takeFirst().at(0).toString(), action);
        }
    }
    
    void testAIActions() {
        SignalBus bus;
        QSignalSpy spy(&bus, &SignalBus::agentUiActionRequested);
        
        QStringList actions = {
            "ai.run_detection", "ai.run_segmentation", "ai.video_tracking",
            "ai.hide_results", "ai.training_model", "ai.view_training_charts"
        };
        
        for (const QString& action : actions) {
            emit bus.agentUiActionRequested(action, QVariantMap());
            QCOMPARE(spy.count(), 1);
            QCOMPARE(spy.takeFirst().at(0).toString(), action);
        }
    }

    void testReconstructionActions() {
        SignalBus bus;
        QSignalSpy spy(&bus, &SignalBus::agentUiActionRequested);
        
        QStringList actions = {
            "reconstruction.load_images", "reconstruction.start_reconstruction",
            "reconstruction.view_3d_model", "reconstruction.close_3d_model"
        };
        
        for (const QString& action : actions) {
            emit bus.agentUiActionRequested(action, QVariantMap());
            QCOMPARE(spy.count(), 1);
            QCOMPARE(spy.takeFirst().at(0).toString(), action);
        }
    }

    void testViewerActions() {
        SignalBus bus;
        QSignalSpy spy(&bus, &SignalBus::agentUiActionRequested);
        
        QStringList actions = {
            "viewer.load_2d", "viewer.load_3d", "viewer.load_dicom"
        };
        
        for (const QString& action : actions) {
            emit bus.agentUiActionRequested(action, QVariantMap());
            QCOMPARE(spy.count(), 1);
            QCOMPARE(spy.takeFirst().at(0).toString(), action);
        }
    }
    
    void testGeneralActions() {
        SignalBus bus;
        QSignalSpy spy(&bus, &SignalBus::agentUiActionRequested);
        
        QVariantMap langParams;
        langParams["language"] = "vi";
        emit bus.agentUiActionRequested("language.change", langParams);
        QCOMPARE(spy.count(), 1);
        QList<QVariant> args = spy.takeFirst();
        QCOMPARE(args.at(0).toString(), QString("language.change"));
        QCOMPARE(args.at(1).toMap()["language"].toString(), QString("vi"));
        
        emit bus.agentUiActionRequested("help.about", QVariantMap());
        QCOMPARE(spy.count(), 1);
        QCOMPARE(spy.takeFirst().at(0).toString(), QString("help.about"));
    }

    void testActionAcknowledgement() {
        SignalBus bus;
        QSignalSpy spy(&bus, &SignalBus::agentUiActionCompleted);

        QVariantMap result;
        result["action"] = "mail.open";
        emit bus.agentUiActionCompleted("request-42", true, result);

        QCOMPARE(spy.count(), 1);
        const QList<QVariant> args = spy.takeFirst();
        QCOMPARE(args.at(0).toString(), QString("request-42"));
        QCOMPARE(args.at(1).toBool(), true);
        QCOMPARE(args.at(2).toMap().value("action").toString(), QString("mail.open"));
    }
};

QTEST_MAIN(TestAgentActions)
#include "TestAgentActions.moc"

File: AIAssistant/test_agent_actions.py
import os
import sys
import unittest

# Add AIAssistant to python path
sys.path.append(os.path.dirname(os.path.abspath(__file__)))
from modules.agent_module import _match_desktop_action, _match_desktop_action_sequence


class TestAgentActions(unittest.TestCase):
    def test_admin_login(self):
        res = _match_desktop_action("dang nhap he thong")
        self.assertEqual(res, {"action": "admin.login", "username": "Admin", "password": "1"})
        
        res2 = _match_desktop_action("please login")
        self.assertEqual(res2, {"action": "admin.login", "username": "Admin", "password": "1"})

    def test_admin_logout(self):
        self.assertEqual(_match_desktop_action("dang xuat"), {"action": "admin.logout"})
        self.assertEqual(_match_desktop_action("logout now"), {"action": "admin.logout"})

    def test_mail_open(self):
        self.assertEqual(_match_desktop_action("mo hop thu cua toi"), {"action": "mail.open"})
        self.assertEqual(_match_desktop_action("open mail"), {"action": "mail.open"})
        self.assertEqual(_match_desktop_action("mo mail"), {"action": "mail.open"})

    def test_assistant_reloads(self):
        self.assertEqual(_match_desktop_action("tai lai model"), {"action": "assistant.reload_model"})
        self.assertEqual(_match_desktop_action("reload rag"), {"action": "assistant.reload_rag"})
        self.assertEqual(_match_desktop_action("tai lai agent"), {"action": "assistant.reload_agent"})
        self.assertEqual(_match_desktop_action("reload server"), {"action": "assistant.reload_server"})
        
    def test_language_change(self):
        self.assertEqual(_match_desktop_action("doi sang english"), {"action": "language.change", "language": "en"})
        self.assertEqual(_match_desktop_action("doi sang tieng viet"), {"action": "language.change", "language": "vi"})

    def test_admin_settings(self):
        self.assertEqual(_match_desktop_action("mo cai dat"), {"action": "admin.settings"})
        self.assertEqual(_match_desktop_action("settings"), {"action": "admin.settings"})

    def test_avatar_and_password(self):
        self.assertEqual(_match_desktop_action("doi avatar"), {"action": "admin.change_avatar"})
        self.assertEqual(_match_desktop_action("change password"), {"action": "admin.change_password"})

    def test_ai_actions(self):
        self.assertEqual(_match_desktop_action("run detection"), {"action": "ai.run_detection"})
        self.assertEqual(_match_desktop_action("phan doan"), {"action": "ai.run_segmentation"})
        self.assertEqual(_match_desktop_action("theo doi video"), {"action": "ai.video_tracking"})
        self.assertEqual(_match_desktop_action("an ket qua"), {"action": "ai.hide_results"})
        self.assertEqual(_match_desktop_action("train model"), {"action": "ai.training_model"})
        self.assertEqual(_match_desktop_action("bieu do training"), {"action": "ai.view_training_charts"})

    def test_reconstruction_actions(self):
        self.assertEqual(_match_desktop_action("bat dau tai tao"), {"action": "reconstruction.start_reconstruction"})
        self.assertEqual(_match_desktop_action("xem 3d model"), {"action": "reconstruction.view_3d_model"})
        self.assertEqual(_match_desktop_action("close 3d model"), {"action": "reconstruction.close_3d_model"})
        self.assertEqual(_match_desktop_action("tai anh tai tao"), {"action": "reconstruction.load_images"})

    def test_reconstruction_load_then_start_workflow(self):
        self.assertEqual(
            _match_desktop_action_sequence(
                "\u0054\u1ea3i \u1ea3nh ch\u1ee5p v\u00e0 t\u00e1i t\u1ea1o 3d."
            ),
            [
                {"action": "reconstruction.load_images"},
                {"action": "reconstruction.start_reconstruction"},
            ],
        )

    def test_viewer_actions(self):
        self.assertEqual(_match_desktop_action("tai 2d"), {"action": "viewer.load_2d"})
        self.assertEqual(_match_desktop_action("tai 3d"), {"action": "viewer.load_3d"})
        self.assertEqual(_match_desktop_action("Tải giúp tôi mô hình 3d."), {"action": "viewer.load_3d"})
        self.assertEqual(_match_desktop_action("load 3d model"), {"action": "viewer.load_3d"})
        self.assertEqual(_match_desktop_action("load dicom"), {"action": "viewer.load_dicom"})

    def test_about(self):
        self.assertEqual(_match_desktop_action("gioi thieu"), {"action": "help.about"})
        self.assertEqual(_match_desktop_action("about"), {"action": "help.about"})

if __name__ == '__main__':
    unittest.main()
